p2p_beef_tx.go

```diff
@@ -39,6 +45,189 @@ type DecodedBEEF struct {
 	ProcessedTxData TxData
 }
 
+func (dBeef *DecodedBEEF) GetMerkleRoots() ([]string, error) {
+	var merkleRoots []string
+	for _, cmp := range dBeef.CMPSlice {
+		partialMerkleRoots, err := cmp.calculateMerkleRoots()
+		if err != nil {
+			return nil, err
+		}
+		merkleRoots = append(merkleRoots, partialMerkleRoots...)
+	}
+	return merkleRoots, nil
+}
+
+// ExecuteSimplifiedPaymentVerification executes the SPV for decoded BEEF tx
+func (dBeef *DecodedBEEF) ExecuteSimplifiedPaymentVerification(provider MerkleRootVerifier) error {
+	err := dBeef.satoshisInInputsGreaterThanZero()
+	if err != nil {
+		return err
+	}
+
+	err = dBeef.satoshisInOutputsGreaterThanZero()
+	if err != nil {
+		return err
+	}
+
+	err = dBeef.validateSatoshisSum()
+	if err != nil {
+		return err
+	}
+
+	err = dBeef.validateLockTime()
+	if err != nil {
+		return err
+	}
+
+	err = dBeef.validateScripts()
+	if err != nil {
+		return err
+	}
+
+	err = dBeef.verifyMerkleRoots(provider)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (dBeef *DecodedBEEF) satoshisInOutputsGreaterThanZero() error {
+	if len(dBeef.ProcessedTxData.Transaction.Outputs) == 0 {
+		return errors.New("invalid output, no outputs")
+	}
+	return nil
+}
+
+func (dBeef *DecodedBEEF) satoshisInInputsGreaterThanZero() error {
+	if len(dBeef.ProcessedTxData.Transaction.Inputs) == 0 {
+		return errors.New("invalid input, no inputs")
+	}
+	return nil
+}
+
+func (dBeef *DecodedBEEF) verifyMerkleRoots(provider MerkleRootVerifier) error {
+	merkleRoots, err := dBeef.GetMerkleRoots()
+	if err != nil {
+		return err
+	}
+
+	res, err := provider.VerifyMerkleRoots(context.Background(), merkleRoots)
+	if err != nil {
+		return err
+	}
+
+	if !res.AllConfirmed {
+		return errors.New("not all merkle roots were confirmed")
+	}
+	return nil
+}
+
+func (dBeef *DecodedBEEF) validateScripts() error {
+	for _, input := range dBeef.ProcessedTxData.Transaction.Inputs {
+		txId := input.PreviousTxID()
+		for j, input2 := range dBeef.InputsTxData {
+			if input2.Transaction.TxID() == string(txId) {
+				result := verifyScripts(dBeef.ProcessedTxData.Transaction, input2.Transaction, j)
+				if !result {
+					return errors.New("invalid script")
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (dBeef *DecodedBEEF) validateSatoshisSum() error {
+	inputSum, outputSum := uint64(0), uint64(0)
+	for i, input := range dBeef.ProcessedTxData.Transaction.Inputs {
+		input2 := dBeef.InputsTxData[i]
```

Comment:
Maybe you could change `input2` to `parentTx` (or inputParentTx??) since it is tx data.

Reply:
Renamed